### PR TITLE
Add clicked position to PlayerUseUnknownEntityEvent

### PR DESCRIPTION
**Pull request by kennytv**
Fixes the event being called 4 times (2 per hand) when a player right clicks with no difference to the event data. Also adds some javadoc and removes an unnecessary access transformer

### CODE REVIEW

File: patches/api/0036-Add-PlayerUseUnknownEntityEvent.patch [hunk 1]
```diff
@@ -3,56 +3,96 @@ From: Jedediah Smith <jedediah@silencegreys.com>
...
 +import org.bukkit.entity.Player;
 +import org.bukkit.event.HandlerList;
 +import org.bukkit.event.player.PlayerEvent;
 +import org.bukkit.inventory.EquipmentSlot;
 +import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
 +
++/**
++ * Represents an event that is called when a player right-clicks an unknown entity.
++ * Useful for plugins dealing with virtual entities (entities that don't actually spawned on the server).
```
Comment (lynxplay) [inline patches/api/0036-Add-PlayerUseUnknownEntityEvent.patch:29]:
```suggestion
+ * Useful for plugins dealing with virtual entities (entities that aren't actually spawned on the server).
```

File: patches/api/0036-Add-PlayerUseUnknownEntityEvent.patch [hunk 2]
```diff
@@ -3,56 +3,96 @@ From: Jedediah Smith <jedediah@silencegreys.com>
...
++    public PlayerUseUnknownEntityEvent(@NotNull Player who, int entityId, boolean attack, @NotNull EquipmentSlot hand, @Nullable Position clickedPosition) {
 +        super(who);
 +        this.entityId = entityId;
 +        this.attack = attack;
 +        this.hand = hand;
++        this.clickedPosition = clickedPosition;
 +    }
 +
++    /**
++     * Returns the location of the unknown entity that was interacted with.
```
Comment (lynxplay) [inline patches/api/0036-Add-PlayerUseUnknownEntityEvent.patch]:
```suggestion
+     * Returns the entity id of the unknown entity that was interacted with.
```

File: patches/server/0083-Add-PlayerUseUnknownEntityEvent.patch [hunk 3]
```diff
@@ -3,63 +3,70 @@ From: Jedediah Smith <jedediah@silencegreys.com>
...
++    private void callPlayerUseUnknownEntityEvent(ServerboundInteractPacket packet, InteractionHand hand, @Nullable net.minecraft.world.phys.Vec3 pos) {
 +        this.cserver.getPluginManager().callEvent(new com.destroystokyo.paper.event.player.PlayerUseUnknownEntityEvent(
 +            this.getCraftPlayer(),
 +            packet.getEntityId(),
-+            packet.getActionType() == ServerboundInteractPacket.ActionType.ATTACK,
-+            hand == InteractionHand.MAIN_HAND ? EquipmentSlot.HAND : EquipmentSlot.OFF_HAND
-+        ));
++            packet.isAttack(),
++            hand == InteractionHand.MAIN_HAND ? EquipmentSlot.HAND : EquipmentSlot.OFF_HAND,
++            pos != null ? io.papermc.paper.math.Position.fine(pos.x, pos.y, pos.z) : null)
```
Comment (lynxplay) [inline patches/server/0083-Add-PlayerUseUnknownEntityEvent.patch]: Should we be adding MCUtil helpers for position creation (from Vec3 and BlockPos respectively) if we want to expand the usage of position here ?

**Review by lynxplay — DISMISSED** [review]

File: patches/server/0009-MC-Utils.patch [hunk 4]
```diff
@@ -3960,6 +3964,14 @@ index 0000000000000000000000000000000000000000..6779a0cc401231c53545dd22827b404b
 +        return new BlockPos(vec.getBlockX(), vec.getBlockY(), vec.getBlockZ());
 +    }
 +
++    public FinePosition toPosition(Vec3 vector) {
++        return Position.fine(vector.x, vector.y, vector.z);
++    }
++
++    public BlockPosition toPosition(Vec3i vector) {
++        return Position.block(vector.getX(), vector.getY(), vector.getZ());
++    }
```
Comment (NoahvdAa) [inline patches/server/0009-MC-Utils.patch]: this doesn't compile

**Review by NoahvdAa — CHANGES_REQUESTED** [review]
Looks good apart from the fact that it doesn't compile

**Comment by kennytv** [comment]
Shhh

**Review by NoahvdAa — APPROVED** [review]

File: patches/api/0036-Add-PlayerUseUnknownEntityEvent.patch [hunk 5]
```diff
@@ -3,56 +3,96 @@ From: Jedediah Smith <jedediah@silencegreys.com>
...
 +    }
 +
++    /**
++     * Returns the exact position that was clicked, or null if not available.
++     * See {@link org.bukkit.event.player.PlayerInteractAtEntityEvent} for more details.
++     *
++     * @return the exact position that was clicked, or null if not available
++     * @see org.bukkit.event.player.PlayerInteractAtEntityEvent
++     */
++    public @Nullable Position getClickedPosition() {
```
Comment (Machine-Maker) [inline patches/api/0036-Add-PlayerUseUnknownEntityEvent.patch]: I think the vector in the packet is a relative vector to the armor stand, it's not a position.

**Review by Machine-Maker — CHANGES_REQUESTED** [review]

**Comment by Cryptite** [comment]
This isn't exactly the right place for this comment, probably, but is there some way in which this event could either be cancelled, or toggled as "consumed" in such a way where interacting with an unknown entity stops all subsequent future actions? And/or maybe there's another way to already handle this. Examples are:

- Right-clicking a packet NPC holding a consumable both interacts with it, but also you start eating.
- Same thing but when holding projectiles (pearls, potions, etc)

**Comment by kennytv** [comment]
It'd be best if you created an issue for that

**Review by Machine-Maker — APPROVED** [review]

**Pull request merged**